README.md
# cljyenc

A Clojure library to encode binary data into YEnc and decode text encoded in YEnc. The libraries name is pronouned (See-El-Jenk)

## Usage

The niche use of YEnc
is to encode binary data into strings that won't confuse text based protocols of network applications.This library takes bytes and encodes them into a string using YEnc. My own usage is to encode bytes into String and back, since the built in Java character encodings aren't able to consistently and accurately reproduce byte streams from input strings in my experience. YEnc was originally used as a replacement for UUEncode on USENET. MIME has superseeded text based protocols for a while. 

## Analysis

The following is a description of the YEnc algorithm. It is a very simple procedure. YEnc relies on using most of the ASCII character set. It's historical use was for encoding USNET messages. We provide the basic algorithm in pseudocode.

### Encoding

The following describes how each character is encoded. All characters are shifted by 42 modulo 256. This is to avoid escaping frequently encountered lengthy chains of NULL or 0 characters. If after shifting we end up with one of the 5 critical characters (NULL, 9, 10, 13, or the escape character "="), we escape it with an = and shift the ordinal of the character by an additional 64. Hence the name *Yenc* as in *Wh**Y** **Enc**ode?* 

```
FOR EACH byte IN byte-array:
    ADD 42 TO byte MOD 256, STORE IN x
    CASE x OF
        0, 9, 10, 13, 61: OUTPUT "=" and x + 64
        DEFAULT: OUTPUT x
```

### Decoding

The encoding is reversed. If we encounter an escape character "=", we subtract 106 (64 + 42) from the following character. Otherwise we subtract 42 from each character to recover the original values.

```
SET critical_flag FALSE
FOR EACH char in character-array:
    STORE ORDINAL VALUE OF char IN x
    CASE x OF
        case 61: 
            read next character as critical (set read-next-as-critical-flag to true), and continue
        DEFAULT:
            IF critical_flag THEN
                SET critical_flag FALSE
                output x - 64 - 42 (or - 106) MOD 256, and continue
            ELSE
                output x - 42 MOD 256, and continue
```

### Related Java Caveats

In Java `byte` is signed, so we must must promote and convert `byte`s to unsigned `int`s with the static method `Byte.toUnsignedInt()` to coerce the right numbers out of  the computations. To change the integer back we use the instance method of `Integer`, `.byteValue()` which returns the unsigned integer to a signed version.

### Known Flaws and Limitations

The algorithm is a first iteration, and could use some optimizations and improved input validation.

## License

Copyright © 2022 Andy Vogel

This program is released under the MIT License. It is free software. This software is distributed as-is with no warranty. See LICENSE for more details.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased] - 2022-04-29
### Changed
- Created project, and copied files to repository.

## [0.1.1] - 2022-04-30
### Changed
- Fixed check for =yend. The code now checks for the epilogue properly.

### Removed
- When features are removed they will be added here.

### Fixed
- Fixes will be added here.


test/cljyenc/core_test.clj
(ns cljyenc.core-test
  (:require [clojure.test :refer :all]
            [cljyenc.core :refer :all]
            [cljyenc.encode :as yenc]
            [cljyenc.decode :as ydec])
  (:import [java.util.random RandomGenerator RandomGeneratorFactory]))

(def test-global-rng-factory (RandomGeneratorFactory/of "L128X1024MixRandom"))
(def test-global-rng (.create test-global-rng-factory))
(def max-string-size 200000)
(def max-file-size 200000)
(def number-of-tests 10000)

(defn yenc-decode-matching-encode? [headers-input headers-output]
  (let [local-rng (.create test-global-rng-factory)
        random-bytes (byte-array max-string-size)]
    (.nextBytes local-rng random-bytes)
    (every? true? (map #(= %1 %2) (seq random-bytes) (ydec/string->bytes (yenc/bytes->string (seq random-bytes) headers-output) headers-input)))))

;; Test to make sure YEnc input = decoded output
(deftest test-yenc-decode-encode
  (testing "Testing YEnc: Every encoded and decoded byte must match the original byte..."
    (is (and
         ;; Should decode fine with or without =ybegin and =yend
         (yenc-decode-matching-encode? true true)
         (yenc-decode-matching-encode? false false)
         ;; Should decode fine for the decode to ignore =ybegin and =yend
         (yenc-decode-matching-encode? false true)))))

;; Test for exception if we don't get a header when we expect one.
(deftest test-yenc-decode-encode-fail-no-header
  (testing "Testing YEnc: Should fail if input doesn't have a header when expecting it..."
    (is (try (yenc-decode-matching-encode? true false) false (catch Exception _ true)))))



src/cljyenc/decode.clj
(ns cljyenc.decode
  (:require [cljyenc.core :refer [yenc-decode-string-to-byte-seq
                                  yenc-find-prologue
                                  yenc-find-epilogue]]))

(defn string->bytes 
  "Converts a YEnc string into a sequence of bytes. 
   s is a string containing Yenc data
   header is true to seek a =ybegin and =yend header
   in the file."
  [s header]
  (let [enc-start  (yenc-find-prologue (seq s))
        found-epi  (yenc-find-epilogue (seq s))]
    (if header (if (and enc-start found-epi)
        ;; Using drop-last to remove extra bytes caused by the =yend epilogue
                 (drop-last 4 (yenc-decode-string-to-byte-seq enc-start))
                 (throw (Exception. "YEnc prologue or epilogue is missing during decoding!")))
        (if found-epi (drop-last 4 (yenc-decode-string-to-byte-seq enc-start))
            (yenc-decode-string-to-byte-seq enc-start)))))

(defn file->bytes 
  "Reads a YEncoded file and reads a sequence of bytes into memory. Always
   assumes yenc files have a header in them"
  [filename]
  (string->bytes (slurp filename) true))

src/cljyenc/core.clj
(ns cljyenc.core
  (:require [clojure.string :as string]))

;; Everything in this file is low level. It should never be called by
;; the user.

(defn yenc-is-critical? [b]
  (or (= b 0)
      (= b 9)
      (= b 10)
      (= b 13)
      (= b 61)))

(defn yenc-find-prologue [char-seq]
  (loop [residue char-seq
         verify-header-mode false]
    (cond (empty? residue) nil
          verify-header-mode (let [possible-header (apply str (take 6 residue))]
                               (if (= possible-header "ybegin")
                                 (drop 6 residue)
                                 (recur (drop 6 residue) false)))
          (= (first residue) \=) (recur (rest residue) true))))


(defn yenc-find-epilogue [char-seq]
  (loop [residue char-seq
         verify-header-mode false]
    (cond (empty? residue) nil
          verify-header-mode (let [possible-header (apply str (take 6 residue))]
                               (if (= possible-header "end")
                                 true (drop 6 residue)))
          (= (first residue) \=) (recur (rest residue) true))))

(defn yenc-encode-byte-to-character [b]
  (let [shifted-character (mod (+ (Byte/toUnsignedInt (byte b)) 42) 256)]
    (if (yenc-is-critical? shifted-character)
      (str \= (char (mod (+ shifted-character 64) 256)))
      (char shifted-character))))

(defn yenc-encode-byte-to-character-seq [byte-seq]
  (seq (apply str (for [b byte-seq] (yenc-encode-byte-to-character b)))))

(defn yenc-decode-character-to-byte-seq [character-seq]
  (loop [remain character-seq
         char-accum (transient [])
         critical false]
    (let [c (first remain)]
      (cond (empty? remain) (persistent! char-accum)
            critical (recur (rest remain) (conj! char-accum (.byteValue (mod (- (int c) 42 64) 256))) false)
            (= c \=) (recur (rest remain) char-accum true)
            :else (recur (rest remain) (conj! char-accum (.byteValue (mod (- (int c) 42) 256))) false)))))

(defn yenc-encode-byte-seq-to-string [bseq]
  (apply str (yenc-encode-byte-to-character-seq bseq)))

(defn yenc-decode-string-to-byte-seq [s]
  (yenc-decode-character-to-byte-seq (seq s)))